taigaio/taiga
# u#1633 PATCH /projects/{id}

![](https://media.giphy.com/media/smZwjAnBYiqQ31Lmt6/giphy-downsized.gif)

This PR have two main features:
- main: adds an endpoint to PATCH /projects/{id}
- secondary: refactors validators to deal with updates in a better way
- poyaque: changed factory.create for factory.build in tests_services

TBD:
- I haven't been able to create a test for a PATCH that deletes the logo, because of how the test client interprets the files
- The new task was created within projects module because:
  - I tried to put it in a more generic place (`base/utils/files`) but found lots of Circular Import errors 😓 
  - I'm not sure if the deletion of the directory (for thumbnails and all) would work _as is_ for other entities as well

How to review:
- [x] check the code. There are two separated commits; better to check them separatedly
- [x] tests are green
- [x] create a project (important note below) and do several of the possible changes and check the response is correct in all tests:
  - change name, description and logo
  - ignore name, change description, ignore logo
  - ignore name, delete description, change logo
  - change name, ignore description, delete logo
  - change name, change description, change logo (again)
  - try to delete name (should get a 400-bad-request)

**important note** If you're using Postman to tests it, you must take into consideration:
- the form has the "logo" field twice
- one to add / change the logo (a file with value)
- the other to delete the loog (a string with "")
- check one XOR the other XOR none (never check both of them)
